python/cudf/cudf/core/dataframe.py

```diff
@@ -7362,6 +7343,16 @@ def equals(self, other):
                 return False
         return super().equals(other)
 
+    def _drop_rows_by_labels(
+        self: "cudf.DataFrame", labels: ColumnLike
+    ) -> "cudf.DataFrame":
+        """Delete rows specified by `label` parameter. In `DataFrame`, this can
+        be achieved efficiently by a left-anti join operation
+
+        labels: a list of labels specifying the rows to drop
+        """
+        return self.join(cudf.DataFrame(index=labels), how="leftanti")
```

Comment:
Nice - very elegant!

Reply:
Since the join parameters are fixed here, it might be worth jumping straight to the cython without first creating a dataframe. Something like

```
idx = Table(index=labels)
res = cudf._lib.join.join(self, idx, 'leftanti', None, left_on=[], right_on=[], left_index=True, right_index=True)
return self.__class__._from_table(res)
```

This way you avoid all the python overhead of constructing a dataframe and then a bunch of unecessary checking and typecasting that the merge codepath uses, which isn't necessarily needed in this case. The downside is you might want to throw if the labels passed dont match the dtype of the target series index.

Reply:
Is there a fast path that could exist in the merge code that we're not currently enabling?

Reply:
Not sure how useful it would be for users, but after the upcoming join improvements it might be useful as an internal utility for `join` at least, since join is just a narrow case of merge with `left_index=True`, `right_index=True` and no `on` parameters. 

A lot of the merge code is there just to validate all the combinations of arguments a user could pass and fix their dtypes if necessary. We _should_ be able to get around that internally if we know something about what our cython API expects. It would also help if `cudf._lib.join.join` had a better "developer facing" API, so the actual call to it from didn't seem so weird. 

The counterargument is that to avoid libcudf itself erroring in an invalid case, we have to do a dtype check, and it could be argued that we might as well let the full user facing merge code error since it's already built to error in that case. But then, if someone passes something invalid, it's almost just as weird that the error comes from inside join.

Reply:
> A lot of the merge code is there just to validate all the combinations of arguments a user could pass and fix their dtypes if necessary. We should be able to get around that internally...

Won't we still need dtype validation, etc., here? The user-provided inputs to `drop()` could be of different dtype from the index columns. 

> It would also help if cudf._lib.join.join had a better "developer facing" API, so the actual call to it from didn't seem so weird.

Agree that we need a better developer-facing join API. FWIW, that won't be coming from Cython though, as the Cython API is going to be very simple (just returning a tuple of gathermaps).

> But then, if someone passes something invalid, it's almost just as weird that the error comes from inside join

Maybe we could wrap the call to `join` here in a `try...except` and raise a more informative error message?

Reply:
> That's exactly my point about the dtypes. If someone passes something that's the wrong type, something inside the merge code will error, is that weird? (EDIT: actually it will just typecast...maybe it SHOULD error, cc @isVoid )

If we want to avoid merge from typecasting, we could expose an internal `merge` API with a `cast=False` kwarg for example. So instead of `lhs.merge(rhs)` we could do `lhs._merge(rhs, cast=False)`.

If we don't want any errors from `merge` to be propagated up to the use, we could catch them and reraise with an appropriate error message. Something like:

```python
def drop(self, labels...):

    try:
         self.index.merge(labels)
    except ValueError:
          raise ValueError("Invalid label type")
```

To be clear, I think it's OK to punt on all these things until after join has been refactored :-)

Reply:
I'm fine with that, I guess my only question is in general, should we avoid using our own user facing API entrypoints internally in favor of a lower level API, just to avoid all the extra overhead we put in the user facing APIs that are meant to check every edge case of everything they could possibly be doing wrong?

Reply:
I think we're in 100%  agreement that we should use internal APIs that bypass Pandas semantics for efficiency. My only point is that doesn't always have to be calling all the way down to Cython. For example [`_gather`](https://github.com/rapidsai/cudf/blob/branch-0.18/python/cudf/cudf/core/frame.py#L590) is an internal API at the Python level.

Here specifically, it looks like we need a `join` API without all the bells and whistles of the top-level `merge()` method, and we currently don't really have that.

Reply:
btw in drop there is this line:
```python3
if errors == "raise" and not labels.isin(levels_index).all():
            raise KeyError("One or more values not found in axis")
```
which I think guards against the dtype mismatch issue?

Reply:
Offline sync with @brandon-b-miller : There is room to optimize for join parameter logic checks, but should defer until the `join` refactor is finished.

Reply:
For now, can we add a TODO here to optimize later by calling a lower-level join API? Ideally, the TODO should document exactly what logic checks we would like to avoid here.